sparta/sparta/kernel/EdgeFactory.hpp

```diff
@@ -0,0 +1,63 @@
+
+// <EdgeFactory> -*- C++ -*-
+
+
+#ifndef __EDGEFACTORY__H__
+#define __EDGEFACTORY__H__
+
+#include "sparta/kernel/Vertex.hpp"
+#include <string>
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include <set>
+
+namespace sparta
+{
+
+/**
+* \class EdgeFactory
+* \brief A class that allows you to create new Vertices
+*
+* A class used to keep track of all the new vertices create
+* and then allow you to 'delete' the vertices once the DAG
+* has been finalized. Used in the DAG when GOPs/Vertices are
+* created to form the DAG, as well as used in Scheduleables
+* to create the internal Vertex used by the Scheduleable to
+* link to the DAG.
+*/
+class EdgeFactory
+{
+
+public:
+
+    //Constructor
+    EdgeFactory() =default;
+    ~EdgeFactory();
+
+    /**
+     * \brief Factory method to create new Vertices
+     */
+    template<typename ...ArgTypes>
+    Edge* newFactoryEdge(ArgTypes&&... args)
+    {
+        Edge*   new_edge = new Edge(std::forward<ArgTypes>(args)...);
+        edges_.emplace(new_edge);
+        return new_edge;
+    }
+
+    void removeEdge(const Edge* e);
+    void dumpToCSV(std::ostream& os) const;
+
+private:
+    // I use bare Edge* here instead of std::unique_ptr<> since I need
+    // to be able to do a set::find() on the Edge address. The
+    // class destructor zaps the Edges
+    std::set<const Edge*> edges_;
```

Comment:
Shouldn't this be a set of `std::unique_ptr`?

Reply:
The thing that's preventing me from doing that (as was done for VertexFactory), is that I want to do a set::find(Edge*). If I were to use std::set(std::unique_ptr<const Edge*>) I apparently need to first convert the Edge* I'm looking for to a unique_ptr<> so that the two will compare. Sooooo, I just use bare pointers, and ~EdgeFactory() will do the deletions. I think there are probably other ways around this, but I chose to do this one out of simplicity.

Reply:
I see.  One way to do this is to have two data structures -- `std::vector` of `std::unique_ptr` and the `set` that holds bare pointers.  But that's just a nit.

Reply:
Yes, I did think of that way around this too. I decided I was starting to go to too much trouble just to avoid a really simple destructor (and a statement in removeEdge). I do see the advantage to unique_ptr<>. but I couldn't find a way to leverage that which didn't require another data structure or passing around Edge id's to look up in a map

Reply:
ok